internal/provider/kubernetes/controller.go

```diff
@@ -165,212 +166,206 @@ func (r *gatewayAPIReconciler) Reconcile(ctx context.Context, _ reconcile.Reques
+				}
+				r.log.Error(err, "failed to process parametersRef for gatewayclass", "name", acceptedGC.Name)
+				return reconcile.Result{}, err
 			}
-			return reconcile.Result{}, err
 		}
 
-		resourceTree.Namespaces = append(resourceTree.Namespaces, namespace)
-	}
+		if resourcesMap[acceptedGC.Name].EnvoyProxy != nil && resourcesMap[acceptedGC.Name].EnvoyProxy.Spec.MergeGateways != nil {
+			r.mergeGateways[acceptedGC.Name] = *resourcesMap[acceptedGC.Name].EnvoyProxy.Spec.MergeGateways
+		}
 
-	// Process the parametersRef of the accepted GatewayClass.
-	if acceptedGC.Spec.ParametersRef != nil && acceptedGC.DeletionTimestamp == nil {
-		if err := r.processParamsRef(ctx, acceptedGC, resourceTree); err != nil {
-			msg := fmt.Sprintf("%s: %v", status.MsgGatewayClassInvalidParams, err)
-			if err := r.updateStatusForGatewayClass(ctx, acceptedGC, false, string(gwapiv1.GatewayClassReasonInvalidParameters), msg); err != nil {
-				r.log.Error(err, "unable to update GatewayClass status")
-			}
-			r.log.Error(err, "failed to process parametersRef for gatewayclass", "name", acceptedGC.Name)
+		if err := r.updateStatusForGatewayClass(ctx, acceptedGC, true, string(gwapiv1.GatewayClassReasonAccepted), status.MsgValidGatewayClass); err != nil {
+			r.log.Error(err, "unable to update GatewayClass status")
 			return reconcile.Result{}, err
 		}
-	}
 
-	if resourceTree.EnvoyProxy != nil && resourceTree.EnvoyProxy.Spec.MergeGateways != nil {
-		r.mergeGateways = *resourceTree.EnvoyProxy.Spec.MergeGateways
-	}
-
-	if err := r.updateStatusForGatewayClass(ctx, acceptedGC, true, string(gwapiv1.GatewayClassReasonAccepted), status.MsgValidGatewayClass); err != nil {
-		r.log.Error(err, "unable to update GatewayClass status")
-		return reconcile.Result{}, err
-	}
+		// The Store is triggered even when there are no Gateways associated to the
+		// GatewayClass. This would happen in case the last Gateway is removed and the
+		// Store will be required to trigger a cleanup of envoy infra resources.
+		r.resources.GatewayAPIResources.Store(string(r.classController), resourcesMap.DeepCopy())
 
-	// Update finalizer on the gateway class based on the resource tree.
-	if len(resourceTree.Gateways) == 0 {
-		r.log.Info("No gateways found for accepted gatewayclass")
+		if len(resourcesMap[acceptedGC.Name].Gateways) == 0 {
+			r.log.Info("No gateways found for accepted gatewayclass")
 
-		// If needed, remove the finalizer from the accepted GatewayClass.
-		if err := r.removeFinalizer(ctx, acceptedGC); err != nil {
-			r.log.Error(err, fmt.Sprintf("failed to remove finalizer from gatewayclass %s",
-				acceptedGC.Name))
-			return reconcile.Result{}, err
-		}
-	} else {
-		// finalize the accepted GatewayClass.
-		if err := r.addFinalizer(ctx, acceptedGC); err != nil {
-			r.log.Error(err, fmt.Sprintf("failed adding finalizer to gatewayclass %s",
-				acceptedGC.Name))
-			return reconcile.Result{}, err
+			// If needed, remove the finalizer from the accepted GatewayClass.
```

Comment:
looks like the order is changed here a bit, shouldnt we update the finalizer and then run `Store`, so any labels show up in the gatewayapi layer and can be used/if required in the future ?

Reply:
I have changed the order, so the store was called when the last gateway was removed and then clean up finalizer. 
I could verify that running the store after the finalizers sometimes resulted in not cleaning up infra resources, e.g. deleting all resources at once.

Reply:
I was wrong, now with the new structure, it should be called afterwards